services/gitlab/gitlab-issues.service.js

```diff
@@ -0,0 +1,291 @@
+        message: '4',
+        color: 'yellow',
+      },
+      documentation: documentation + labelDocumentation,
+    },
+    {
+      title: 'GitLab all issues',
+      pattern: 'all/:project+',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: { gitlab_url: 'https://gitlab.com' },
+      staticPreview: {
+        label: 'issues',
+        message: '115k all',
+        color: 'yellow',
+      },
+      documentation,
+    },
+    {
+      title: 'GitLab all issues',
+      pattern: 'all-raw/:project+',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: { gitlab_url: 'https://gitlab.com' },
+      staticPreview: {
+        label: 'all issues',
+        message: '115k',
+        color: 'yellow',
+      },
+      documentation,
+    },
+    {
+      title: 'GitLab all issues by-label',
+      pattern: 'all-raw/:project+',
+      namedParams: {
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: {
+        labels: 'test,failure::new',
+        gitlab_url: 'https://gitlab.com',
+      },
+      staticPreview: {
+        label: 'all test,failure::new issues',
+        message: '20',
+        color: 'yellow',
+      },
+      documentation: documentation + labelDocumentation,
+    },
+  ]
+
+  static defaultBadgeData = { label: 'issues', color: 'informational' }
+
+  static render({ issues }) {
+    return {
+      label: issues.renderLabel,
+      message: issues.renderMsg,
+      color: issues.renderColor,
+    }
+  }
+
+  async fetch({ project, baseUrl, labels }) {
+    // https://docs.gitlab.com/ee/api/issues_statistics.html#get-project-issues-statistics
+    return super.fetch({
+      schema,
+      url: `${baseUrl}/api/v4/projects/${encodeURIComponent(
+        project
+      )}/issues_statistics`,
+      options: labels ? { searchParams: { labels } } : undefined,
+      errorMessages: {
+        404: 'project not found',
+      },
+    })
+  }
+
+  static transform({ variant, raw, statistics, labels }) {
+    const state = variant
+    const isMultiLabel = labels && labels.includes(',')
+    const labelText = labels ? `${isMultiLabel ? `${labels}` : labels} ` : ''
+
+    let labelPrefix = ''
+    let messageSuffix = ''
+    if (raw !== undefined) {
+      labelPrefix = `${state} `
+    } else {
+      messageSuffix = state
+    }
+
+    let issueCount
+    switch (state) {
+      case 'open':
+        issueCount = statistics.counts.opened
+        break
+      case 'closed':
+        issueCount = statistics.counts.closed
+        break
+      case 'all':
+        issueCount = statistics.counts.all
+        break
+    }
+    const renderLabel = `${labelPrefix}${labelText}issues`
+    const renderMsg = `${metric(issueCount)}${
+      messageSuffix ? ' ' : ''
+    }${messageSuffix}`
+    const renderColor = issueCount > 0 ? 'yellow' : 'brightgreen'
+
+    return {
+      renderLabel,
+      renderMsg,
+      renderColor,
+    }
+  }
```

Comment:
Appreciate you making the updates, but you've actually swung too far in the opposite direction. There's some more detail in [this section](https://github.com/badges/shields/blob/master/doc/TUTORIAL.md#43-querying-an-api) of our tutorial, but our service classes all follow a very a standard flow (they are all essentially doing the same thing after all!) and accordingly utilize a standard set of functions. We do this for several reasons, including readability, testability, and separation of concerns amongst others.

Some classes are simple enough that they don't need to utilize every function from that set of standard functions, but the set of functions includes:

- `handle`: entry point and orchestrator
- `fetch`: retrieve raw response from upstream endpoint
- `transform`: process raw response and extract/transform needed data point(s)
- `render`: specify the final badge attributes, like label, message, and color.

The line between `transform` and `render` _can_ get a little blurry in some extreme cases, but generally speaking, the `render` function shouldn't be processing the raw response and it should strictly be focused on setting the label, etc.. Similarly, the `transform` function should only be processing the raw response and getting the result(s), which will then be passed along to the `render` function because it will use that value to determine what to specify for the badge message, color, etc.

In this case, I think a `transform` function is probably a good idea to avoid having to put too much work into the `handle` function. Such a `transform` function should be responsible for determining the issue count (based on the variant) and it also needs to account for the case where the `statistics` field is `null` in the response. That's all it should do, and it should simple return back the count of issues to the caller.

Reply:
Thanks for your guidance. I'm actually not very good at JS, but your explanation and official documentation helped me a lot.

Reply:
Glad it helped. And I never would've known you didn't have much experience with JS if you hadn't mentioned it, so perhaps you're better than you think!